### PR TITLE
Add ErrorKind::OutOfMemory

### PR DESCRIPTION
**Pull request by kornelski**
Ability to express `ENOMEM` as an `io::Error`.

I've used `OutOfMemory` as opposed to `NotEnoughMem` or `AllocationFailed`, because "OOM" is used in Rust already.

See also #84612

### CODE REVIEW

**Comment by JohnTitor** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/84785#issuecomment-830594074
We should tweak this clippy test: https://github.com/rust-lang/rust/blob/5a4ab26459a1ccf17ef5bb4c841d3ae5517b2890/src/tools/clippy/tests/ui/wildcard_enum_match_arm.rs

@bors r-

**Comment by kornelski** [comment]
@JohnTitor I've updated the Clippy test

**Comment by JohnTitor** [comment]
```
2021-05-02T19:45:52.1008217Z Can't find any online and idle self-hosted runner in current repository that matches the required labels: 'ubuntu-latest'
2021-05-02T19:45:52.1387762Z Can't find any online and idle self-hosted runner in current repository's organization account that matches the required labels: 'ubuntu-latest'
2021-05-02T19:45:52.2775891Z Found online and idle hosted runner in current repository's organization account that matches the required labels: 'ubuntu-latest'
```

@bors retry

**Comment by dtolnay** [comment]
@rfcbot fcp merge

I guess JFYI @rust-lang/libs.

**Comment by joshtriplett** [comment]
The idea of adding more error codes did come up a few times, but you're right, an fcp is warranted to have all of that discussion in public and make sure everyone gets polled.

@rfcbot merge

**Comment by m-ou-se** [comment]
> EDIT: hmmm, rfcbot doesn't seem to be responding...

@joshtriplett I don't think it watches merged/closed PRs. An FCP would require a new issue.

File: library/std/src/io/error.rs [hunk 1]
```diff
@@ -186,6 +186,11 @@ pub enum ErrorKind {
     /// This means that the operation can never succeed.
     #[stable(feature = "unsupported_error", since = "1.53.0")]
     Unsupported,
+
+    /// An operation could not be completed, because it failed
+    /// to allocate enough memory.
+    #[stable(feature = "out_of_memory_error", since = "1.53.0")]
```
[the remark this answers is not in the document]
Comment (kornelski) [reply library/std/src/io/error.rs:192]: #84856

**Comment by joshtriplett** [comment]
That was my mistake, sorry; this was discussed verbally but not FCPed, and despite the positive outcome of that discussion, it does still need an FCP to check for (and document) consensus. I'll start a new issue to do that FCP poll.

**Pull request merged**